src/System.Collections/src/System/Collections/Generic/SortedList.cs

```diff
@@ -152,10 +152,31 @@ public SortedList(IDictionary<TKey, TValue> dictionary, IComparer<TKey> comparer
             if (dictionary == null)
                 throw new ArgumentNullException(nameof(dictionary));
 
-            dictionary.Keys.CopyTo(_keys, 0);
-            dictionary.Values.CopyTo(_values, 0);
-            Array.Sort<TKey, TValue>(_keys, _values, comparer);
-            _size = dictionary.Count;
+            int count = dictionary.Count;
+            if (count != 0)
+            {
+                dictionary.Keys.CopyTo(_keys, 0);
+                dictionary.Values.CopyTo(_values, 0);
+                Debug.Assert(count == _keys.Length);
+                if (count > 1)
+                {
+                    Array.Sort<TKey, TValue>(_keys, _values, comparer);
+                    comparer = Comparer; // obtain default if this is null.
+                    TKey keyPrev = _keys[0];
+                    for (int i = 1; i != _keys.Length; ++i)
+                    {
+                        TKey keyNext = _keys[i];
+                        if (comparer.Compare(keyPrev, keyNext) >= 0)
+                        {
+                            throw new ArgumentException(SR.Format(SR.Argument_AddingDuplicate, keyNext));
+                        }
+
+                        keyPrev = keyNext;
```

Comment:
No strong reason, aside from I seem to be developing an instinct for pulling array elements local if they are used more than once. Again though, I think comparing `_keys[i - 1]` and `_keys[i]` will be clearer as to the intent, so I'll change. I might pull `_keys` itself local though, rather than have n-1 field accesses.